In [1]:
%load_ext autoreload
%autoreload 2

import sys
from pathlib import Path
sys.path.append('../')

import pandas as pd
import os
import subprocess
import zipfile
import matplotlib.pyplot as plt
import lightgbm as lgb
import numpy as np

from itertools import product
from sklearn.model_selection import GroupShuffleSplit
from itertools import chain
from sklearn.model_selection import GroupKFold
from src.utils import * 
from src.feature_engineering import *
from src.pipeline import *
from src.run import *
from src.encoding import *

In [2]:
train_df = basic_data_pipeline()

Train shape: (18145372, 126)
Mem. usage decreased to 13843.82 Mb (19.2% reduction)


In [3]:
mean_score, std_score, scores = get_cv_score(train_df, n_splits=3)

/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 1: Hit Rate@3=0.4639, AUC=0.8121


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 2: Hit Rate@3=0.4662, AUC=0.8040


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 3: Hit Rate@3=0.4654, AUC=0.8056
Mean: 0.4652 ± 0.0009


------------
## Feature Engineering 
------------

- 기본적인 데이터 변환 마무리: frequentFlyer

In [4]:
train_df = fe1_pipeline(train_df)

In [5]:
pd.reset_option('display.max_columns')
pd.set_option('display.max_columns', None)

df.apply(함수, axis=1)  
axis=1: 각 행에 대해 함수 적용, axis=0: 각 열에 대해 함수 적용

In [6]:
get_cv_score(train_df, n_splits=3)

/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 1: Hit Rate@3=0.4668, AUC=0.8146


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 2: Hit Rate@3=0.4648, AUC=0.8097


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 3: Hit Rate@3=0.4668, AUC=0.8048
Mean: 0.4661 ± 0.0009


(np.float64(0.4660988064588458),
 np.float64(0.0009467497071236656),
 [0.4667791116843602, 0.4647599591419816, 0.46675734855019574])

In [9]:
train_df = is_direct_flight(train_df)

In [10]:
get_cv_score(train_df, n_splits=3)

/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 1: Hit Rate@3=0.4668, AUC=0.8146


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 2: Hit Rate@3=0.4648, AUC=0.8097


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 3: Hit Rate@3=0.4668, AUC=0.8048
Mean: 0.4661 ± 0.0009


(np.float64(0.4660988064588458),
 np.float64(0.0009467497071236656),
 [0.4667791116843602, 0.4647599591419816, 0.46675734855019574])

### legs_segments_ 관련 데이터 먼저 fe 진행 
  1. Economics (cabinClass, baggage)  
	legs0_segments0_baggageAllowance_quantity  
	legs0_segments0_baggageAllowance_weightMeasurementType  
	legs0_segments0_cabinClass  

  2. Operations (aircraft, carrier)  
	legs0_segments0_flightNumber  
	legs0_segments0_aircraft_code  
	legs0_segments0_marketingCarrier_code  
	legs0_segments0_operatingCarrier_code  

  3. Geography (airport)  
	legs0_segments0_arrivalTo_airport_city_iata 
	legs0_segments0_arrivalTo_airport_iata  
	legs0_segments0_departureFrom_airport_iata  

  4. Availability (seats)  
	legs0_segments0_seatsAvailable  

In [11]:
train_test = baggage_measurement_type(train_df)
get_cv_score(train_test, n_splits=3)

/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/feature_engineering.py:118: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[f'legs{i}_segments{j}_is_no_baggage'] = (df[weight_col] == 1) & (df[quantity_col] == 0)
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/feature_engineering.py:117: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[f'legs{i}_segments{j}_is_weight_based'] = (df[weight_col] == 1) & (df[quantity_col] >= 3)
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/feature_engineering.py:11

Fold 1: Hit Rate@3=0.4668, AUC=0.8146


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 2: Hit Rate@3=0.4648, AUC=0.8097


/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:75: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  train_groups = train_df.groupby('ranker_id').size().values
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:76: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  val_groups = val_df.groupby('ranker_id').size().values # df/series -> numpy
/Users/jaewoo/Desktop/Kaggle/pred_aeroclub/notebook/../src/run.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=Tr

Fold 3: Hit Rate@3=0.4668, AUC=0.8048
Mean: 0.4661 ± 0.0009


(np.float64(0.4660988064588458),
 np.float64(0.0009467497071236656),
 [0.4667791116843602, 0.4647599591419816, 0.46675734855019574])

In [14]:
for col in train_test.columns:
    if 'baggage' in col:
        print(f"{col}: {train_test[col].sum()}")

legs0_segments0_baggageAllowance_quantity: 47804112.0
legs0_segments0_baggageAllowance_weightMeasurementType: 1070127.0
legs0_segments1_baggageAllowance_quantity: 30667160.0
legs0_segments1_baggageAllowance_weightMeasurementType: 907343.0
legs0_segments2_baggageAllowance_quantity: 2299192.0
legs0_segments2_baggageAllowance_weightMeasurementType: 74572.0
legs1_segments0_baggageAllowance_quantity: 37036152.0
legs1_segments0_baggageAllowance_weightMeasurementType: 852181.0
legs1_segments1_baggageAllowance_quantity: 24400988.0
legs1_segments1_baggageAllowance_weightMeasurementType: 726590.0
legs1_segments2_baggageAllowance_quantity: 1672582.0
legs1_segments2_baggageAllowance_weightMeasurementType: 55263.0
legs0_segments0_is_no_baggage: 57578
legs0_segments1_is_no_baggage: 45254
legs0_segments2_is_no_baggage: 5130
legs1_segments0_is_no_baggage: 47225
legs1_segments1_is_no_baggage: 34984
legs1_segments2_is_no_baggage: 3601


In [15]:
for col in [col for col in train_test.columns if 'baggage' in col]:
    print(f"{col}: variance = {train_test[col].var()}")

legs0_segments0_baggageAllowance_quantity: variance = 50.47740936279297
legs0_segments0_baggageAllowance_weightMeasurementType: variance = 0.05199046432971954
legs0_segments1_baggageAllowance_quantity: variance = 186.5028076171875
legs0_segments1_baggageAllowance_weightMeasurementType: variance = 0.18520726263523102
legs0_segments2_baggageAllowance_quantity: variance = 216.48031616210938
legs0_segments2_baggageAllowance_weightMeasurementType: variance = 0.23442697525024414
legs1_segments0_baggageAllowance_quantity: variance = 53.378074645996094
legs1_segments0_baggageAllowance_weightMeasurementType: variance = 0.05553501099348068
legs1_segments1_baggageAllowance_quantity: variance = 193.48326110839844
legs1_segments1_baggageAllowance_weightMeasurementType: variance = 0.19114616513252258
legs1_segments2_baggageAllowance_quantity: variance = 224.32601928710938
legs1_segments2_baggageAllowance_weightMeasurementType: variance = 0.24422693252563477
legs0_segments0_is_no_baggage: variance = 

train_df['legs0_segments0_baggageAllowance_weightMeasurementType'] == 1  
train_df['legs0_segments0_baggageAllowance_quantity'] >= 3  
두 가지의 교집합 -> 무게 방식으로 짐을 측정하는 케이스  

그러나, 아래와 같은 케이스에서 57578개의 예외 케이스 발생  
train_df['legs0_segments0_baggageAllowance_weightMeasurementType'] == 1  
train_df['legs0_segments0_baggageAllowance_quantity'] == 0  
MeasurementType == 1이 무게를 측정하는 방식이라면, 수하물을 0개 수용하는게 말이 안됨  
-> 다른 측정 방식이 존재하거나 이 데이터가 의미하는 바가 있을 것  

이를 어떻게 알아볼 수 있나?  
-> 일단 유사한 다른 데이터 참고하여 조사  
-> 1. 다른 segment에서도 동일한 일 발생하는지 -> segment1에서도 동일하게 발생  
-> 2. 특정 항공사에서 발생하는 일인지  

***
발견한 특성  
1. cabinClass가 99퍼센트 이상 1일 때 해당 케이스 발현  
-> 저가 항공사의 방편이거나 값싼 항공편의 특징인 것 같음
2. FV, SU, S7 등 기존 많이 이용된 항공사에서는 이런 케이스가 없었고 TK, EY 두 개의 항공사가 99퍼센트 이상을 차지  
3. 또 한가지 특징적인게, 전체 데이터의 평균 비행시간은 4시간인데 이 둘은 10시간이 넘어감 이건 굉장히 큰 차이  
4. 여태까지 정보를 종합해보면, 이런 케이스는 낮은 등급의 좌석이고 특정 항공사일 때 나타남  

-> No baggage feature을 새로 만들기  
-> measurementType == 1인 경우를 no_baggage와 weight_type으로 분리  
-> 피처를 만들긴 했지만 < 1퍼센트로 너무 값이 작아 무의미  
-> 이렇게 작은 값일 땐 통계적으로 무의미해진다는 사실 기억, 내일은 cabin class부터 접근 